src/ir_Kelon.cpp

```diff
@@ -0,0 +1,500 @@
+void IRKelonAc::setRaw(const uint64_t new_code) {
+  _.raw = new_code;
+}
+
+/// Convert a standard A/C mode (stdAc::opmode_t) into it a native mode.
+/// @param[in] mode A stdAc::opmode_t operation mode.
+/// @return The native mode equivalent.
+uint8_t IRKelonAc::convertMode(const stdAc::opmode_t mode) {
+  switch (mode) {
+    case stdAc::opmode_t::kCool:
+      return kKelonModeCool;
+    case stdAc::opmode_t::kHeat:
+      return kKelonModeHeat;
+    case stdAc::opmode_t::kDry:
+      return kKelonModeDry;
+    case stdAc::opmode_t::kFan:
+      return kKelonModeFan;
+    default:
+      return kKelonModeSmart;
+  }
+}
+
+/// Convert a standard A/C fan speed (stdAc::fanspeed_t) into it a native speed.
+/// @param[in] fan A stdAc::fanspeed_t fan speed
+/// @return The native speed equivalent.
+uint8_t IRKelonAc::convertFan(stdAc::fanspeed_t fan) {
+  switch (fan) {
+    case stdAc::fanspeed_t::kMin:
+    case stdAc::fanspeed_t::kLow:
+      return kKelonFanMin;
+    case stdAc::fanspeed_t::kMedium:
+      return kKelonFanMedium;
+    case stdAc::fanspeed_t::kHigh:
+    case stdAc::fanspeed_t::kMax:
+      return kKelonFanMax;
+    default:
+      return kKelonFanAuto;
+  }
+}
+
+/// Convert a native mode to it's stdAc::opmode_t equivalent.
+/// @param[in] mode A native operating mode value.
+/// @return The stdAc::opmode_t equivalent.
+stdAc::opmode_t IRKelonAc::toCommonMode(const uint8_t mode) {
+  switch (mode) {
+    case kKelonModeCool:
+      return stdAc::opmode_t::kCool;
+    case kKelonModeHeat:
+      return stdAc::opmode_t::kHeat;
+    case kKelonModeDry:
+      return stdAc::opmode_t::kDry;
+    case kKelonModeFan:
+      return stdAc::opmode_t::kFan;
+    default:
+      return stdAc::opmode_t::kAuto;
+  }
+}
+
+/// Convert a native fan speed to it's stdAc::fanspeed_t equivalent.
+/// @param[in] speed A native fan speed value.
+/// @return The stdAc::fanspeed_t equivalent.
+stdAc::fanspeed_t IRKelonAc::toCommonFanSpeed(const uint8_t speed) {
+  switch (speed) {
+    case kKelonFanMin:
+      return stdAc::fanspeed_t::kLow;
+    case kKelonFanMedium:
+      return stdAc::fanspeed_t::kMedium;
+    case kKelonFanMax:
+      return stdAc::fanspeed_t::kHigh;
+    default:
+      return stdAc::fanspeed_t::kAuto;
+  }
+}
+
+/// Convert the internal A/C object state to it's stdAc::state_t equivalent.
+/// @return A stdAc::state_t containing the current settings.
+stdAc::state_t IRKelonAc::toCommon() const {
+  stdAc::state_t result{};
+  result.protocol = decode_type_t::KELON;
+  result.model = -1;  // Unused.
+  result.mode = toCommonMode(getMode());
+  result.celsius = true;
+  result.degrees = getTemp();
+  result.fanspeed = toCommonFanSpeed(getFan());
+  result.turbo = getSupercool();
+  result.sleep = getSleep() ? 0 : -1;
+  // Not supported.
+  result.power = true;  // N/A, AC only supports toggling it
+  result.swingv = stdAc::swingv_t::kAuto;  // N/A, AC only supports toggling it
```

Comment:
You could/should allow passing in a optional `prev` state and checking/reporting the correct response after checking for the toggle.

See: Airwell, Coolix, Daikin(64|128), Midea, PanasonicAc32, ToshibaAC, TranscoldAc, Voltas, & WhirlpoolAc for examples.